server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/futurerelease/FutureReleaseMailParameterHook.java

```diff
@@ -0,0 +1,110 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FutureReleaseMailParameterHook implements MailParametersHook {
+    private static final int UTC_TIMESTAMP_LENGTH = 20;
+    private static final int ZONED_TIMESTAMP_LENGTH = 25;
+    private static final Logger LOGGER = LoggerFactory.getLogger(FutureReleaseMailParameterHook.class);
+
+    public static final ProtocolSession.AttachmentKey<FutureReleaseParameters.HoldFor> FUTURERELEASE_HOLDFOR = ProtocolSession.AttachmentKey.of("FUTURERELEASE_HOLDFOR", FutureReleaseParameters.HoldFor.class);
+
+    private final Clock clock;
+
+    @Inject
+    public FutureReleaseMailParameterHook(Clock clock) {
+        this.clock = clock;
+    }
+
+    @Override
+    public HookResult doMailParameter(SMTPSession session, String paramName, String paramValue) {
+        try {
+            Duration requestedHoldFor = evaluateHoldFor(paramName, paramValue);
+
+            if (requestedHoldFor.compareTo(MAX_HOLD_FOR_SUPPORTED) > 0) {
+                LOGGER.debug("HoldFor is greater than max-future-release-interval or holdUntil exceeded max-future-release-date-time");
+                return HookResult.DENY;
+            }
+            if (requestedHoldFor.isNegative()) {
+                LOGGER.debug("HoldFor value is negative or holdUntil value is before now");
+                return HookResult.DENY;
+            }
+            if (session.getAttachment(FUTURERELEASE_HOLDFOR, Transaction).isPresent()) {
+                LOGGER.debug("Mail parameter cannot contains both holdFor and holdUntil parameters");
+                return HookResult.DENY;
+            }
+            session.setAttachment(FUTURERELEASE_HOLDFOR, FutureReleaseParameters.HoldFor.of(requestedHoldFor), Transaction);
+            return HookResult.DECLINED;
+        } catch (IllegalArgumentException e) {
+            LOGGER.debug("Incorrect syntax when handling FUTURE-RELEASE mail parameter", e);
+            return HookResult.DENY;
+        }
+    }
+
+    private Duration evaluateHoldFor(String paramName, String paramValue) {
+        if (paramName.equals(HOLDFOR_PARAMETER)) {
+            return Duration.ofSeconds(Long.parseLong(paramValue));
+        }
+        if (paramName.equals(HOLDUNTIL_PARAMETER)) {
+            DateTimeFormatter formatter = selectDateTimeFormatter(paramValue);
+            Instant now = LocalDateTime.now(clock).toInstant(ZoneOffset.UTC);
+            return Duration.between(now, ZonedDateTime.parse(paramValue, formatter).toInstant());
+        }
+        throw new IllegalArgumentException("Invalid parameter name " + paramName);
+    }
+
+    private DateTimeFormatter selectDateTimeFormatter(String dateTime) {
+        if (dateTime.length() == UTC_TIMESTAMP_LENGTH) {
+            return DateTimeFormatter.ISO_INSTANT.withZone(ZoneId.of("Z"));
+        }
+        if (dateTime.length() == ZONED_TIMESTAMP_LENGTH) {
+            return DateTimeFormatter.ISO_OFFSET_DATE_TIME;
+        }
```

Comment:
Use `optionalStart` to come up with a dedicated `DateTimeFormatter`

Reply:
There is no optional part in this formatter. How do I use `optionalStart`?

Reply:
I guess Benoit means to use `ImapDateTimeFormatter.rfc5322()`.